### PR TITLE
Chore: System Tests for Lesson Completion

### LINKED ISSUE #1865: Chore: System Tests for Completing Lessons
### Description
System test coverage for the completing lessons feature.

### Acceptance Criteria
- [ ] Marking a lesson complete on the lesson page
- [ ] Marking a lesson incomplete on the lesson page
- [ ] Marking a lesson complete on the course page
- [ ] Marking a lesson incomplete on the course page

### PR DESCRIPTION
**Pull request by rlmoser99**
Because
* Lesson Completion needs end to end test coverage.

This commit:
* Tests lesson completion and incompletion on lesson page.
* Tests lesson completion and incompletion on course page.
* Resolves #1865 

I am still rather new to writing system tests, so I have two questions.

I followed the pattern that was established in the lesson_completion_spec file, but found the wording of the test and the expectations to be inconsistent. For example, we are testing that a user can complete a lesson and the test is asserting that the incomplete_btn is not nil. I understand that this is what visually happens, but I was under the impression that the test statement should more closely mirror the assertion. Is this normal/expected, or should I re-word the test statements?

In the course_lesson_completion_spec file, I followed the same pattern to test when a user is not signed in. However, the only way that I could figure out how to test it, is by testing the lack of button using the class of the lesson completion buttons. Since testing the lack of an element could give us false positives, is there another way to test this that I am missing? 

EDIT: I forgot about running rubocop. Before submitting, I double checked the contributing guide to make sure I wasn't missing anything. It does not list running `rubocop`, only `rspec`. Should we update this? 

I will submit these rubocop changes with any additional changes that will be needed.

### CODE REVIEW

**Comment by KevinMulhern** [comment]
> I followed the pattern that was established in the lesson_completion_spec file, but found the wording of the test and the expectations to be inconsistent. For example, we are testing that a user can complete a lesson and the test is asserting that the incomplete_btn is not nil. I understand that this is what visually happens, but I was under the impression that the test statement should more closely mirror the assertion. Is this normal/expected, or should I re-word the test statements?

Yeah this is a good point, I think we need more assertions to test this behaviour maybe? 

```
expect(page).to have_button(<complete_button>)
expect(page).to have_no_button(<incomplete_button>)
expect(user.lesson_completions.lesson_ids).to include(lesson.id)
```

and the inverse of for the incomplete state.

> In the course_lesson_completion_spec file, I followed the same pattern to test when a user is not signed in. However, the only way that I could figure out how to test it, is by testing the lack of button using the class of the lesson completion buttons. Since testing the lack of an element could give us false positives, is there another way to test this that I am missing?

I think you've done it the best way 💯  thats the behaviour on the page so makes total sense.

**Comment by rlmoser99** [comment]
I could not use `find(:test_id, 'complete_btn')` to test when a link/button was not present, because it would raise an error `(Capybara::ElementNotFound)`. I did not want to use different syntax to test the two different situations, so I decided to use `have_selector('button[data-test-id="lesson_incomplete_btn"]')` since it worked in both test situations.

**Comment by KevinMulhern** [comment]
Yeah testing for existance is where `find` falls down unforuntley. What about using something like 
`expect(page).to have_no_button('button[data-test-id="lesson_incomplete_btn")`

It just makes it clearer that we're looking for buttons than using `have_selector`.

**Comment by rlmoser99** [comment]
I see your point to make it clear that we are looking for buttons. However, `expect(page).to have_no_button('button[data-test-id="lesson_incomplete_btn"]')` is invalid syntax.

I had previously tried `expect(page).to have_button(test_id: 'lesson_incomplete_btn')`. However, this produces a false positive, because `expect(page).to have_button(test_id: 'any_string')` passes too. This must be a limit from Capybara custom selector.

In addition, the buttons on the lesson page are actually links, so even if the above solution worked, it would not be as readable for those tests. I had also thought about `expect(page).to have_no_link('Complete', title: 'Mark lesson complete')` and `expect(page).to have_link('', title: 'Mark lesson incomplete')`, but this add more implementation details. Plus the course's completion buttons are icons and do not have content, classes, or anything that communicates why we are testing for the presence of them.

So, unless I am missing something, the best solution that I can think of is to create RSpec custom matchers to improve the readability. Since `.to have_no_selector` is preferred over `not_to have_selector`, I created two different matchers. In addition, I tested similar logic to above false positive and it fails as expected `expect(page).to have_submit_button('any_string')`.

File: spec/support/custom_matchers.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+RSpec::Matchers.define :have_submit_button do |value|
```
Comment (KevinMulhern) [inline spec/support/custom_matchers.rb:1]: Would it be possible to add these custom matchers as capybara selectors instead? like we've done with the test id?

**Review by KevinMulhern — COMMENTED** [review]
> I see your point to make it clear that we are looking for buttons. However, expect(page).to have_no_button('button[data-test-id="lesson_incomplete_btn"]') is invalid syntax.

sorry about that, I typed that out as pseudo code without testing it.

> the buttons on the lesson page are actually links, so even if the above solution worked, it would not be as readable for those tests.

Those _should_ be buttons, we pushed out a [fix to a bug fairly recently](https://github.com/TheOdinProject/theodinproject/pull/1858) that changes them from links to buttons. You might have to update your branch to fetch the latest changes.

I really like the custom matcher solution you've come up with @rlmoser99, it makes those tests so much easier to follow. I think it would be worth trying to add them as capybara selectors, just because thats the only context they will be used in.

Other than that, This is looking fantastic 💯

File: spec/support/custom_matchers.rb (hunk 1 above)
Comment (rlmoser99) [reply spec/support/custom_matchers.rb]: I may be missing something, as I do not have much experience customizing Capybara. However, based on the research that I did, it seemed to make the most sense to add helper methods that utilize the custom capybara selector that we already had. Take a look at the new commit and let me know if I should have done this differently.

**Comment by rlmoser99** [comment]
@KevinMulhern Thanks for letting me know about the recent changes. I did have them already. The lesson pages are the ones with links instead of buttons:
https://github.com/TheOdinProject/theodinproject/blob/master/app/views/lessons/_lesson_completion_state.html.erb

**Review by KevinMulhern — APPROVED** [review]
Looks great to me @rlmoser99, really nice work! 👍

**Pull request merged**